ambry-api/src/main/java/com/github/ambry/clustermap/VcrClusterAgentFactory.java

```diff
@@ -18,12 +18,15 @@
 
 
 /**
- * A factory to create {@link ClusterSpectator} object.
+ * A factory interface to get {@link VcrClusterParticipant} and {@link VcrClusterSpectator}.
  */
-public class HelixClusterSpectatorFactory implements ClusterSpectatorFactory {
+public interface VcrClusterAgentFactory {
```

Comment:
Should this. implement `ClusterAgentsFactory`? I was wondering if this class shouldn't return ClusterPartiticipant and ClusterSpectator at all. Instead of that it can have methods like -- getPeerCloudReplica(), getAssignedPartition(), getPeerStoreReplicas() etc. Things like `VcrClusterParticipant` and `VcrClusterSpectator` would then become internal implementation details, and vcr replication logic wouldn't need to know how to deal with participant and spectator.

Reply:
Basically, you are trying to model this as a `clustermap` and the code in this PR is trying to model this by roles. 

It's nice to have a unified interface to provide getAssignedPartition(), getPeerStoreReplicas() etc. 
However, `getAssignedPartition()` only belongs to `Participant`; getPeerReplica() can have two implementation, one for `Spectator` and the other for Participant`.  With this in mind, we need to provide role registration method in this `"clustermap"`.  This increases the implementation complexity. 

The code in this PR split functionalities by role. `Participant` provides `getAssignedPartition()`, `getPeerReplica()`;  `Spectator`  provides `getPeerReplica(replica)`. It's more natural. In addition, this can give node flexibility to be what role(s).

From abstraction wise, it's good to do sth like you suggested. But from implementation wise, it's clear to have two roles, each has their functionalities.

Reply:
but anyway, I will change VcrClusterSpectator to not implement helix interfaces and only deal with ambry defined listener interfaces.  (Maybe next PR)